### PR TITLE
[new release] server-reason-react (0.1.0)

### PR DESCRIPTION
**Pull request by davesnx**
Rendering React components on the server natively

- Project page: <a href="https://github.com/ml-in-barcelona/server-reason-react">https://github.com/ml-in-barcelona/server-reason-react</a>

## 0.1.0

Initial release of server-reason-react, includes:

- Server-side rendering of ReasonReact components (renderToString, renderToStaticMarkup & renderToLwtStream)
- `server-reason-react.browser_ppx` for skipping code from the server
- `server-reason-react.melange_ppx` for enabling melange bindings and extensions which run on the server
- `server-reason-react.belt` a native Belt implementation
- `server-reason-react.js` a native Js implementation (unsafe and limited)
- `server-reason-react.url` and `server-reason-react.url-native` a universal library with both implementations to work with URLs on the server and the client
- `server-reason-react.promise` and `server-reason-react.promise-native` a universal library with both implementations to work with Promises on the server and the client. Based on https://github.com/aantron/promise
- `server-reason-react.melange-fetch` a fork of melange-fetch which is a melange library to fetch data on the client via the Fetch API. This fork is to be able to compile it on the server (not running).
- `server-reason-react.webapi` a fork of melange-webapi which is a melange library to work with the Web API on the client. This fork is to be able to compile it on the server (not running).

### CODE REVIEW

File: packages/server-reason-react/server-reason-react.0.1.0/opam [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+maintainer: ["David Sancho <dsnxmoreno@gmail.com>"]
+authors: ["David Sancho <dsnxmoreno@gmail.com>"]
+license: "MIT"
+homepage: "https://github.com/ml-in-barcelona/server-reason-react"
+bug-reports: "https://github.com/ml-in-barcelona/server-reason-react/issues"
+depends: [
+  "dune" {>= "3.8"}
+  "ocaml" {>= "5.0.0"}
+  "reason" {>= "3.10.0"}
+  "melange"
```
Comment (mseri) [inline packages/server-reason-react/server-reason-react.0.1.0/opam:12]:
```suggestion
  "melange" {>= "2.0.0"}
```
It does not compile with melange 1.0, let's try this.

Comment (mseri) [reply packages/server-reason-react/server-reason-react.0.1.0/opam:12]: For
```
#=== ERROR while compiling server-reason-react.0.1.0 ==========================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/ml-in-barcelona/server-reason-react/releases/download/0.1.0/server-reason-react-0.1.0.tbz)
# path                 ~/.opam/5.0/.opam-switch/build/server-reason-react.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p server-reason-react -j 71 @install
# exit-code            1
# env-file             ~/.opam/log/server-reason-react-7-e7d174.env
# output-file          ~/.opam/log/server-reason-react-7-e7d174.out
### output ###
# File "packages/promise/js/dune", line 7, characters 25-35:
# 7 |  (libraries melange.belt melange.js)
#                              ^^^^^^^^^^
# Error: Library "melange.js" not found.
# -> required by library "server-reason-react.promise-js" in
#    _build/default/packages/promise/js
# -> required by _build/default/META.server-reason-react
# -> required by _build/install/default/lib/server-reason-react/META
# -> required by _build/default/server-reason-react.install
# -> required by alias install
# File "packages/url/js/dune", line 5, characters 12-22:
# 5 |  (libraries melange.js)
#                 ^^^^^^^^^^
# Error: Library "melange.js" not found.
# -> required by library "server-reason-react.url-js" in
#    _build/default/packages/url/js
# -> required by _build/default/META.server-reason-react
# -> required by _build/install/default/lib/server-reason-react/META
# -> required by _build/default/server-reason-react.install
# -> required by alias install
```

**Comment by mseri** [comment]
I don't know what is needed for this:
```

#=== ERROR while compiling server-reason-react.0.1.0 ==========================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-base-compiler.5.1.1 | pinned(https://github.com/ml-in-barcelona/server-reason-react/releases/download/0.1.0/server-reason-react-0.1.0.tbz)
# path                 ~/.opam/5.1/.opam-switch/build/server-reason-react.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p server-reason-react -j 255 @install
# exit-code            1
# env-file             ~/.opam/log/server-reason-react-7-c2f9aa.env
# output-file          ~/.opam/log/server-reason-react-7-c2f9aa.out
### output ###
# File "demo/client/dune", line 5, characters 30-42:
# 5 |  (libraries melange shared_js reason-react)
#                                   ^^^^^^^^^^^^
# Error: Library "reason-react" not found.
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlopt.opt -w -40 -w -49 -g -I packages/melange.js/.js.objs/byte -I packages/melange.js/.js.objs/native -I /home/opam/.opam/5.1/lib/lwt -I /home/opam/.opam/5.1/lib/pcre -intf-suffix .ml -no-alias-deps -o packages/melange.js/.js.objs/native/js.cmx -c -impl packages/melange.js/Js.pp.ml)
# File "packages/melange.js/Js.ml", line 526, characters 17-30:
# 526 |     | Pcre.Error WorkspaceSize -> raise @@ Invalid_argument "WorkspaceSize"
#                        ^^^^^^^^^^^^^
# Error: This variant pattern is expected to have type Pcre.error
#        There is no constructor WorkspaceSize within type Pcre.error
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc.opt -w -40 -w -49 -g -bin-annot -I packages/melange.js/.js.objs/byte -I /home/opam/.opam/5.1/lib/lwt -I /home/opam/.opam/5.1/lib/pcre -intf-suffix .ml -no-alias-deps -o packages/melange.js/.js.objs/byte/js.cmo -c -impl packages/melange.js/Js.pp.ml)
# File "packages/melange.js/Js.ml", line 526, characters 17-30:
# 526 |     | Pcre.Error WorkspaceSize -> raise @@ Invalid_argument "WorkspaceSize"
#                        ^^^^^^^^^^^^^
# Error: This variant pattern is expected to have type Pcre.error
#        There is no constructor WorkspaceSize within type Pcre.error
```
Maybe it requires melange 3.0.0 after all?

**Comment by davesnx** [comment]
I would need to tweak the opam/dune, since I don't want to build the demo on the CI, and also might require melange 3.0.0 as well.

Will push with the fixes

**Comment by davesnx** [comment]
Disabled the demo via profile=dev and added "new-doc"

**Comment by davesnx** [comment]
It's failing without many info and I'm unsure what I need to do to get fixed or I have done something wrong. Any pointers @mseri?

File: packages/server-reason-react/server-reason-react.0.1.0/opam [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    "-p"
+    name
+    "-j"
+    jobs
+    "@install"
+    "@runtest" {with-test}
+    "@new-doc" {with-doc}
+  ]
+]
+url {
```
Comment (mseri) [inline packages/server-reason-react/server-reason-react.0.1.0/opam:43]:
```suggestion
flags: verbose
url {
```
This should show us what is going on

Comment (davesnx) [reply packages/server-reason-react/server-reason-react.0.1.0/opam:44]: The build failed in fsb in 5.1, but still didn't gave us many info. The only special part (from my pov) on this repo is either profile or a dependency that breaks silently (pcre?)

File: packages/server-reason-react/server-reason-react.0.1.0/opam [hunk 3]
```diff
@@ -0,0 +1,52 @@
+opam-version: "2.0"
+synopsis: "Rendering React components on the server natively"
+maintainer: ["David Sancho <dsnxmoreno@gmail.com>"]
+authors: ["David Sancho <dsnxmoreno@gmail.com>"]
+license: "MIT"
+homepage: "https://github.com/ml-in-barcelona/server-reason-react"
+bug-reports: "https://github.com/ml-in-barcelona/server-reason-react/issues"
+depends: [
+  "dune" {>= "3.8"}
```
Comment (mseri) [inline packages/server-reason-react/server-reason-react.0.1.0/opam:9]:
```suggestion
  "dune" {>= "3.9"}
```
Should fix
```

#=== ERROR while compiling server-reason-react.0.1.0 ==========================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-base-compiler.5.1.1 | pinned(https://github.com/ml-in-barcelona/server-reason-react/releases/download/0.1.0/server-reason-react-0.1.0.tbz)
# path                 ~/.opam/5.1/.opam-switch/build/server-reason-react.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p server-reason-react -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/server-reason-react-7-64979c.env
# output-file          ~/.opam/log/server-reason-react-7-64979c.out
### output ###
# File "demo/client/dune", line 4, characters 2-12:
# 4 |  (enabled_if
#       ^^^^^^^^^^
# Error: Unknown field enabled_if
```

**Comment by mseri** [comment]
The comparison failures in the tests may be due to internal details on how `\r\n` is read and trimmed

**Comment by mseri** [comment]
@davesnx it seems that the only remaining failure is
```
# --- a/_build/.sandbox/68c037da8b6acb05f8bb72dc150f4600/default/packages/server-reason-react-ppx/cram/external.t/run.t
# +++ b/_build/.sandbox/68c037da8b6acb05f8bb72dc150f4600/default/packages/server-reason-react-ppx/cram/external.t/run.t.corrected
# @@ -1,5 +1,6 @@
#    $ ../ppx.sh --output re input.re
#    module MyPropIsOptionOptionBoolWithSig = {
# -    [%ocaml.error
# -     "externals aren't supported on server-reason-react. externals are used to bind to React components defined in JavaScript, in the server, that doesn't make sense. If you need to render this on the server, implement a placeholder or an empty element"];
# +    [%%ocaml.error
# +      "externals aren't supported on server-reason-react. externals are used to bind to React components defined in JavaScript, in the server, that doesn't make sense. If you need to render this on the server, implement a placeholder or an empty element"
# +    ];
#    };
```
the difference is a % character and a space. Would it be possible to update (or disable) the cram test and force push the release? Otherwise I can disable the tests, but then you lose metadata fixes on updates of the dependencies

**Comment by davesnx** [comment]
Yes, will update the tests.

File: packages/server-reason-react/server-reason-react.0.1.0/opam [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+license: "MIT"
+homepage: "https://github.com/ml-in-barcelona/server-reason-react"
+bug-reports: "https://github.com/ml-in-barcelona/server-reason-react/issues"
+depends: [
+  "dune" {>= "3.9"}
+  "ocaml" {>= "5.0.0"}
+  "reason" {>= "3.10.0"}
+  "melange" {>= "3.0.0"}
+  "ppxlib" {> "0.23.0" & <= "0.31.0"}
+  "pcre" {<= "7.5.0"}
```
[the remark this answers is not in the document]
Comment (mseri) [reply packages/server-reason-react/server-reason-react.0.1.0/opam:14]: That should already be in `pcre` opam file, so nothing to do from your side in theory

File: packages/server-reason-react/server-reason-react.0.1.0/opam [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+license: "MIT"
+homepage: "https://github.com/ml-in-barcelona/server-reason-react"
+bug-reports: "https://github.com/ml-in-barcelona/server-reason-react/issues"
+depends: [
+  "dune" {>= "3.9"}
+  "ocaml" {>= "5.0.0"}
+  "reason" {>= "3.10.0"}
+  "melange" {>= "3.0.0"}
+  "ppxlib" {> "0.23.0"}
+  "pcre" {<= "7.5.0"}
```
Comment (mseri) [inline packages/server-reason-react/server-reason-react.0.1.0/opam:14]:
```suggestion
  "pcre" {>= "7.4.0"}
```

Comment (mseri) [reply packages/server-reason-react/server-reason-react.0.1.0/opam]: This fixes
```
#=== ERROR while compiling server-reason-react.0.1.0 ==========================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-base-compiler.5.1.1 | pinned(https://github.com/ml-in-barcelona/server-reason-react/releases/download/0.1.0/server-reason-react-0.1.0.tbz)
# path                 ~/.opam/5.1/.opam-switch/build/server-reason-react.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p server-reason-react -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/server-reason-react-7-ea8ede.env
# output-file          ~/.opam/log/server-reason-react-7-ea8ede.out
### output ###
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlopt.opt -w -40 -w -49 -g -I packages/melange.js/.js.objs/byte -I packages/melange.js/.js.objs/native -I /home/opam/.opam/5.1/lib/lwt -I /home/opam/.opam/5.1/lib/pcre -intf-suffix .ml -no-alias-deps -o packages/melange.js/.js.objs/native/js.cmx -c -impl packages/melange.js/Js.pp.ml)
# File "packages/melange.js/Js.ml", line 526, characters 17-30:
# 526 |     | Pcre.Error WorkspaceSize -> raise @@ Invalid_argument "WorkspaceSize"
#                        ^^^^^^^^^^^^^
# Error: This variant pattern is expected to have type Pcre.error
#        There is no constructor WorkspaceSize within type Pcre.error
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc.opt -w -40 -w -49 -g -bin-annot -I packages/melange.js/.js.objs/byte -I /home/opam/.opam/5.1/lib/lwt -I /home/opam/.opam/5.1/lib/pcre -intf-suffix .ml -no-alias-deps -o packages/melange.js/.js.objs/byte/js.cmo -c -impl packages/melange.js/Js.pp.ml)
# File "packages/melange.js/Js.ml", line 526, characters 17-30:
# 526 |     | Pcre.Error WorkspaceSize -> raise @@ Invalid_argument "WorkspaceSize"
#                        ^^^^^^^^^^^^^
# Error: This variant pattern is expected to have type Pcre.error
#        There is no constructor WorkspaceSize within type Pcre.error
```

The upper bound is not necessary. If a new version of PCRE breaks your library we will add the upper bound at that stage when the revdeps are tested

**Comment by mseri** [comment]
I don't know why on some systems we see this failure:
```
# File "packages/melange.ppx/tests/mel_send_pipe.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/d9c7ba773101092e98c25c29d564a3d3/default/packages/melange.ppx/tests/mel_send_pipe.t _build/.sandbox/d9c7ba773101092e98c25c29d564a3d3/default/packages/melange.ppx/tests/mel_send_pipe.t.corrected
# diff --git a/_build/.sandbox/d9c7ba773101092e98c25c29d564a3d3/default/packages/melange.ppx/tests/mel_send_pipe.t b/_build/.sandbox/d9c7ba773101092e98c25c29d564a3d3/default/packages/melange.ppx/tests/mel_send_pipe.t.corrected
# index d1c827a..6747136 100644
# --- a/_build/.sandbox/d9c7ba773101092e98c25c29d564a3d3/default/packages/melange.ppx/tests/mel_send_pipe.t
# +++ b/_build/.sandbox/d9c7ba773101092e98c25c29d564a3d3/default/packages/melange.ppx/tests/mel_send_pipe.t.corrected
# @@ -58,6 +58,7 @@ Make sure is placed correctly
#  Single argument (Ptyp_constr)
#    $ cat > input.ml << EOF
#    > external arrayBuffer : arrayBuffer Js.Promise.t = "arrayBuffer" [@@mel.send.pipe: T.t]
# +  warning: here-document at line 1 delimited by end-of-file (wanted `EOF')
#  
#    $ ./standalone.exe -impl input.ml | ocamlformat - --enable-outside-detected-project --impl | tee output.ml
#    let (arrayBuffer : arrayBuffer Js.Promise.t -> T.t) =
# @@ -77,6 +78,7 @@ Labelled arguments
#    $ cat > input.ml << EOF
#    > type t
#    > external scale : x:float -> y:float -> unit = "scale"[@@mel.send.pipe : t]
# +  warning: here-document at line 1 delimited by end-of-file (wanted `EOF')
#  
#    $ ./standalone.exe -impl input.ml | ocamlformat - --enable-outside-detected-project --impl | tee output.ml
#    type t
# @@ -104,6 +106,7 @@ Nonlabelled arguments as functions
#    $ cat > input.ml << EOF
#    > type t
#    > external forEach : (string -> int -> unit) -> unit = "forEach" [@@mel.send.pipe : t]
# +  warning: here-document at line 1 delimited by end-of-file (wanted `EOF')
#  
#    $ ./standalone.exe -impl input.ml | ocamlformat - --enable-outside-detected-project --impl | tee output.ml
#    type t
# @@ -130,6 +133,7 @@ Nonlabelled arguments as functions
#  'a
#    $ cat > input.ml << EOF
#    > external postMessage : 'a -> string -> unit = "postMessage" [@@mel.send]
# +  warning: here-document at line 1 delimited by end-of-file (wanted `EOF')
#  
#    $ ./standalone.exe -impl input.ml | ocamlformat - --enable-outside-detected-project --impl | tee output.ml
#    let (postMessage : 'a -> string -> unit) =
# @@ -155,6 +159,7 @@ Send pipe with 'a
#    $ cat > input.ml << EOF
#    > type t_window
#    > external postMessage : 'a -> string -> unit = "postMessage" [@@mel.send.pipe : t_window]
# +  warning: here-document at line 1 delimited by end-of-file (wanted `EOF')
#  
#    $ ./standalone.exe -impl input.ml | ocamlformat - --enable-outside-detected-project --impl | tee output.ml
#    type t_window
# (cd _build/default/packages/melange.js && ./test.exe)
# Testing `Js'.
# This run has ID `YVSOF3EO'.
```

**Comment by mseri** [comment]
But it does not prevent the merge to me

**Comment by mseri** [comment]
Thanks a lot

**Comment by davesnx** [comment]
Thanks to you @mseri

**Pull request merged**